enhancements/installer/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,193 @@
+be used as OpenShift Compute Nodes, when they are made available as Nova
+flavors.
+
+## Motivation
+
+In heavy-loaded clusters, or when the workload is better handled with specially
+purposed hardware, the functional requirement of controlling where the work is
+computed might outweight the added value of virtualisation.
+
+In OpenStack clusters where Nova transparently provisions baremetal machines,
+this enhancement will ensure that OpenShift installations can use them as
+Compute Nodes.
+
+### Goals
+
+* Enable running OpenShift on a mix of virtual and baremetal machines.
+* Baremetal machines can be added and removed from a running OpenShift cluster.
+
+### Non-Goals
+
+**Separate subnets.**
+
+An OpenStack setting where baremetal machines can't be attached to a virtual
+subnet is not supported in this enhancement.
+
+**External connectivity via floating IPs.**
+
+The OpenStack cluster might make the baremetal nodes accessible, but exposing
+them via floating IPs might represent an increase in complexity and is
+currently not a goal.
+
+## Proposal
+
+This enhancement is less about adding a new logical piece, than it is about
+uncovering unknowns in a configuration that might work out-of-the-box in
+selected environments.
+
+Therefore, the implementation consists of:
+* identifying a reference architecture
+* anticipating probable issues
+* testing and fixing errors
+
+### Characteristics of the reference architecture
+
+* The choice between deploying a virtual or baremetal machine is done by selecting an appropriate Nova flavour
+  * which can be done both in install config and passed to Cluster API provider OpenStack
+  * for the time being, the interactive prompt does not ask the flavor for the Compute Nodes specifically; the choice must be made editing `install-config.yaml`
+* The deployment request makes the same API call to Nova as with a virtual machine install (there are no explicit calls to Ironic or anything in the Installer required)
+  * so creating baremetal nodes should work transparently from the installer as well as Cluster API provider OpenStack
+* both virtual and baremetal machines consume the same image from glance
+
+### Probable issues
+
+* When attaching and detaching, the expected timings must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+* When attaching, the Certificate Signing Request timeout must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+### User Stories
+
+#### Increased performance
+
+User wants to run the Compute Nodes on baremetal machines for increased
+performance. The Control Plane sits on regular Nova instances to leverage the
+flexibility of virtual machines.
+
+### Risks and Mitigations
+
+**Uncertainty about OpenStack's extra networking resources**
+
+Are ports (including trunk), security groups, allowed address pairs all these necessary for baremetal?
```

Comment:
Security groups are only supported where the switch supports groking and applying the security group. Few and far between with baremetal and ML2 configurations.

Ports are required, and if there is a specific network that would need to be supplied or chosen by the user to "attach" the baremetal machine to, although that also may not be possible in the environment.

Ironic has a concept for representing trunk ports and passing that configuration through, but not all ML2 drivers which plug into neutron support the concept of configuring a trunk. It might be something best left until later.